contracts/eth/nearbridge/contracts/NearBridge.sol

```diff
@@ -202,161 +150,140 @@ contract NearBridge is INearBridge, AdminControlled {
-            if (lastValidAt != 0) {
-                commitBlock();
-            }
-        } else {
+        if (block.timestamp < lastValidAt) {
             require(
-                nearBlock.inner_lite.timestamp >= untrustedHead.timestamp.add(replaceDuration),
-                "NearBridge: can only replace with a sufficiently newer block"
+                nearBlock.inner_lite.timestamp >= untrustedTimestamp.add(replaceDuration),
+                "Can only replace with a sufficiently newer block"
             );
+        } else if (lastValidAt != 0) {
+            curHeight = untrustedHeight;
+            if (untrustedNextEpoch) {
+                curEpoch = (curEpoch + 1) % 3;
+            }
+            lastValidAt = 0;
+
+            blockHashes_[curHeight] = untrustedHash;
+            blockMerkleRoots_[curHeight] = untrustedMerkleRoot;
         }
 
         // Check that the new block's height is greater than the current one's.
-        require(nearBlock.inner_lite.height > head.height, "NearBridge: Height of the block is not valid");
+        require(nearBlock.inner_lite.height > curHeight, "New block must have higher height");
 
         // Check that the new block is from the same epoch as the current one, or from the next one.
-        bool nearBlockIsFromNextEpoch;
-        if (nearBlock.inner_lite.epoch_id == head.epochId) {
-            nearBlockIsFromNextEpoch = false;
-        } else if (nearBlock.inner_lite.epoch_id == head.nextEpochId) {
-            nearBlockIsFromNextEpoch = true;
+        bool fromNextEpoch;
+        if (nearBlock.inner_lite.epoch_id == epochs[curEpoch].epochId) {
+            fromNextEpoch = false;
+        } else if (nearBlock.inner_lite.epoch_id == epochs[(curEpoch + 1) % 3].epochId) {
+            fromNextEpoch = true;
         } else {
-            revert("NearBridge: Epoch id of the block is not valid");
+            revert("Epoch id of the block is not valid");
         }
 
         // Check that the new block is signed by more than 2/3 of the validators.
-        _checkBp(nearBlock, nearBlockIsFromNextEpoch ? nextBlockProducers : currentBlockProducers);
+        Epoch storage thisEpoch = epochs[fromNextEpoch ? (curEpoch + 1) % 3 : curEpoch];
+        // Last block in the epoch might contain extra approvals that light client can ignore.
+        require(nearBlock.approvals_after_next.length >= thisEpoch.numBPs, "Approval list is too short");
+        // The sum of uint128 values cannot overflow.
+        uint256 votedFor = 0;
+        for ((uint i, uint cnt) = (0, thisEpoch.numBPs); i != cnt; ++i) {
+            bytes32 stakes = thisEpoch.packedStakes[i >> 1];
+            if (nearBlock.approvals_after_next[i].some) {
+                votedFor += uint128(bytes16(stakes));
+            }
+            if (++i == cnt) {
+                break;
+            }
+            if (nearBlock.approvals_after_next[i].some) {
+                votedFor += uint128(uint256(stakes));
```

Comment:
I'm trying to figure out still if this is correct. My thought why it might be incorrect:

stakes is 32 bytes: [firstStake (16 bytes), secondStake (16 bytes)]

it is read as a 256bit integer **big-endian** (is it read as big endian?) so first stake is stored in the least significant bits.

After casting to 128, we are left again with `firstStake` rather than with `secondStake`.

Reply:
The current implementation is correct.